third_task.py

```diff
@@ -0,0 +1,224 @@
+                    level=logging.INFO,
+                    format='%(asctime)s:%(levelname)s:%(name)s:%(message)s')
+
+
+def explore_data(url_, params, filename):
+    """
+
+    :param url_: url to download data about 5k user
+    :param params: params to url
+    :param filename: name of our file
+    """
+    logging.info('Exploring the data provider site  and download a csv file with users ')
+    response = requests.get(url_, params=params)
+    if response.status_code == 200:
+        data = response.text
+        with open(filename + '.csv', 'w', newline='', encoding='utf-8') as file:
+            file.write(data)
+
+
+def read_data(filename):
+    """
+
+    :param filename: name of our file
+    :return: explored data(from given file) to our dict to work with
+    """
+    logging.info('reading explored data(from given file) to our dict to work with')
+
+    with open(f'{filename}.csv', 'r', encoding='utf-8') as file:
+        csv_reader = csv.DictReader(file)
+        data = list(csv_reader)
+        # for row in csv_reader:
+        #     data.append(row)
+    return data
+
+
+def give_filtered_data(filename, gender_=None, rows_=None):
+    """
+    :param filename: name of our file
+    :param gender_: on of optional params given from console
+    :param rows_: on of optional params given from console
+    :return: return filtered data explored from our file
+    """
+    logging.info(' return filtered data explored from our file')
+    data = read_data(filename)
+    if gender_:
+        filtered_data = list(filter(lambda x: x['gender'] == gender_, data))
+        return filtered_data
+    elif rows_:
+        filtered_data = []
+        counter = 0
+        for f_data in data:
+            if counter < rows_:
+                filtered_data.append(f_data)
+                counter += 1
+            else:
+                break
+        return filtered_data
+    else:
+        return 'need params gender or rows'
+
+
+def add_fields_to_file(file, rows):
+    """
+    :param rows: ew fields to our file
+     :param file: name of our file
+    :return rows
+    """
+    logging.info('add new fields to our file')
+    for i, row in enumerate(rows, start=1):
+        row['global_index'] = i
+        row['current_time'] = datetime.now()
+
+        match row['name.title']:
+            case 'Mrs':
+                row['name.title'] = 'missis'
+            case 'Ms':
+                row['name.title'] = 'miss'
+            case 'Mr':
+                row['name.title'] = 'mister'
+            case 'Madame':
+                row['name.title'] = 'mademoiselle'
+
+        row['dob.date'] = datetime.strptime(row['dob.date'], '%Y-%m-%dT%H:%M:%S.%fZ').strftime('%m/%d/%Y')
+        row['registered.date'] = datetime.strptime(row['registered.date'], '%Y-%m-%dT%H:%M:%S.%fZ').strftime('%m-%d-%Y/%H:%M:%S')
+
+    with open(file + '.csv', 'w', newline='', encoding='utf-8') as file:
+        csv_writer = csv.DictWriter(file)
```

Comment:
here you overwrite function parameter **file** (which holds filename) to variable **file** which will contain write stream.
be careful